### PR TITLE
Setup Monorepo

### PR DESCRIPTION
**Pull request by xg-wang**
## What it does

This PR sets up the monorepo for several repositories under `ember-fastboot` org, and internal test packages for easier testing across several repos:

- `packages/fastboot`: a low-level tool for anything relying on rendering Ember app in node environment (e.g., custom fastboot-server, prember)
- `packages/ember-cli-fastboot`: required for all SSR ember app
- `packages/fastboot-express-middleware`: low-level tool for building custom fastboot-server
- `packages/fastboot-app-server`: sane default for serving production app
- `test-packages/*`: internal ember apps for acceptance & e2e testing, not published

All packages to be published are bumped to `3.1.2`, which is the largest among the packages' versions.

Also converting several `fastboot` repo tests that rely on `ember-cli-fastboot` for building ember app into `test-packages/integration-tests/test/basic-test.js`.

`ember-cli-fastboot` is already using `release-it-yarn-workspace` to release `ember-cli-fastboot`; we can continue to use the release flow to release others.

## Context

[`ember-fastboot` org](https://github.com/ember-fastboot) has multiple packages useful for server-side rendered ember apps.
Although these packages can be used together with different versions combination (e.g., https://github.com/ember-fastboot/fastboot/blob/v3.1.2/src/fastboot-schema.js is explicitly designed for compatibility), it is usually desired to keep them in sync from maintainers' perspective and use a combination of versions known good from users'.

Currently, when bumping one package, several PRs need to cascade for other packages; it becomes even trickier when a fastboot contributor needs to make a change or write a test that interacts with different packages.

To merge these related packages into a monorepo, we can make maintaining and contributing to fastboot much easier.

The cons for doing monorepo is losing the flexibility to release different versions for each package and increasing the coupling, but they are not necessarily bad IMO.

## Checklist

The monorepo setup needs to ensure a couple of things:
- [x] link packages in dev/test
- [x] ci pass
- [x] move some fastboot tests to integration-tests for demonstration ([basic-test.js](https://github.com/xg-wang/ember-cli-fastboot/blob/monorepo/test-packages/integration-tests/test/basic-test.js))
- [x] update packages/*/readme.md
- [x] release setup
- [x] keep all the git history of the different packages
- (need to verify after merge) make sure the list of contributors on GitHub is a superset

## Next steps

- [ ] After this PR is merged, move issue/pr from other repos then archive (We can only move manually/individually now 😥)
- [ ] Identify and fix any incompatibilities
- [ ] Identify and document remaining duplicated tests from packages, then move them to `test-packages`
- [ ] greenkeeper -> dependapot
- [ ] unify linting
- [ ] fix ci (unskip windows 😥)
- [ ] drop node 13 (fastboot is specifying `"node": "10.* || >=12"` whereas others are specifying `"node": "10.* || 12.* || >= 14"`
- [ ] Merge https://github.com/ember-fastboot/fastboot-website into the monorepo for live demo, prod smoke testing and documentation.

### CODE REVIEW

**Comment by xg-wang** [comment]
I squashed the commits, this commit contains the new test setup [`5451050` (#805)](https://github.com/ember-fastboot/ember-cli-fastboot/pull/805/commits/5451050e349db4f665ead42a44474a79eede7658#diff-e74916b53dfb900af44d77c046b30865b56e58f6673e20b80507098f93fa0989)

**Comment by xg-wang** [comment]
@rwjblue (friendly ping) can I get a review on this PR?

File: README.md [hunk 1]
```diff
@@ -3,8 +3,8 @@
 [![npm version](https://badge.fury.io/js/ember-cli-fastboot.svg)](https://badge.fury.io/js/ember-cli-fastboot)
 [![Actions Status](https://github.com/ember-fastboot/ember-cli-fastboot/workflows/CI/badge.svg)](https://github.com/ember-fastboot/ember-cli-fastboot/actions)
 
-An Ember CLI addon that allows you to render and serve Ember.js apps on
-the server. Using FastBoot, you can serve rendered HTML to browsers and
+FastBoot allows you to render and serve Ember.js apps on the server.
```
Comment (kiwiupover) [inline README.md:6]: This is great.

File: test-packages/custom-sandbox-app/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# custom-sandbox-app
```
Comment (kiwiupover) [inline test-packages/custom-sandbox-app/README.md:1]: @xg-wang what is the difference between this test app and the basic-app?

If there is a difference and you update the read me there to be clear about what this test-package app is doing and what tests should be added to this package. 

You could use this as an example
https://github.com/ember-fastboot/ember-cli-fastboot/tree/master/test-packages

File: test-packages/hot-swap-app/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+# hot-swap-app
+
+This README outlines the details of collaborating on this Ember application.
```
Comment (kiwiupover) [inline test-packages/hot-swap-app/README.md:3]: @xg-wang same question here what is the difference between this test app and the basic-app?

If there is a difference and you update the read me there to be clear about what this test-package app is doing and what tests should be added to this package. 

You could use this as an example https://github.com/ember-fastboot/ember-cli-fastboot/tree/master/test-packages

File: packages/fastboot/test/fastboot-test.js [hunk 4]
```diff
@@ -50,41 +50,6 @@ describe('FastBoot', function() {
     expect(fn).to.throw(/(.+)package.json was malformed or did not contain a fastboot config/);
   });
 
-  it('can render HTML', function() {
```
Comment (kiwiupover) [inline packages/fastboot/test/fastboot-test.js]: @xg-wang this test could be added to the `basic-app` test-package. 

I worry that is we have a lot of different test-packages then developers will not know where to add new tests.

File: packages/fastboot/test/fastboot-test.js [hunk 5]
```diff
@@ -99,21 +64,6 @@ describe('FastBoot', function() {
       });
   });
 
-  it('cannot not render app HTML with shouldRender set as false', function() {
```
Comment (kiwiupover) [inline packages/fastboot/test/fastboot-test.js]: This test can also go in the basic-app

File: packages/fastboot/test/fastboot-test.js [hunk 6]
```diff
@@ -168,33 +118,6 @@ describe('FastBoot', function() {
       });
   });
 
-  it('can serialize the head and body', function() {
```
Comment (kiwiupover) [inline packages/fastboot/test/fastboot-test.js]: Same here this can go in the basic-app

File: packages/fastboot/test/fastboot-test.js [hunk 7]
```diff
@@ -264,56 +187,6 @@ describe('FastBoot', function() {
       });
   });
 
-  it('can reload the distPath', function() {
```
Comment (kiwiupover) [inline packages/fastboot/test/fastboot-test.js]: Move to -> basic-app

File: test-packages/integration-tests/test/basic-test.js [hunk 8]
```diff
@@ -0,0 +1,189 @@
+"use strict";
```
Comment (kiwiupover) [inline test-packages/integration-tests/test/basic-test.js:1]: @xg-wang Do you think we could move these tests to the basic-app?

**Review by kiwiupover — COMMENTED** [review]
@xg-wang I have left a few comments describing which test packages the test should live in.

**Comment by kiwiupover** [comment]
@xg-wang the first few of your commits show a blank diff any idea why?

https://github.com/ember-fastboot/ember-cli-fastboot/pull/805/commits/0f5b00b8442bcc6c08fc2f69d4cfdb710e11eb3a

**Comment by xg-wang** [comment]
@kiwiupover
> this test could be added to the basic-app test-package.
Thanks for pointing out! I assume you are saying the tests I moved from `fastboot/test/fastboot-test.js` to `test-packages/integration-tests/test/basic-test.js`, where fixture `basic-app` is used.
Sorry I forgot to mention explicitly the migration path, which I'm thinking about to move `basic-app/test/*.js` to `test-packages/integration-tests/test`.
The thought here is we use `test-packages/basic-app` purely as a fixture app, and write tests under `test-packages/test/*.js`. I think the advantages are:
- we can test against multiple dists
- we shift the mental model about integration test boundaries. In `basic-app/test` we serve the app and test HTML response, in new setup we use `fastboot.visit` directly, and test middleware (currently `packages/fastboot-express-middleware/test/middleware-test.js` in another file (I haven't created yet). In new middleware integration test we don't need to test each test app, we just need to use some test app (basic-app should be enough most time) to cover middleware behavior
- Not starting a server for every test file should execute faster

I prefer to move `test-packages/basic-app/test/*-test.js` to `test-packages/integration-tests/test/<maybe-new-name>-test.js`. It would be a simple move change at first, and change starting server to use `fastboot.visit` when we have time (hope we _will_ clean up the TODO 😄), as we migrate more old tests in `ember-cli-fastboot` we don't need to start the server. I couldn't think of cons for doing so, except the churn of moving them.

Any thoughts?

---
For the PR description TODO item
>  Identify and document remaining duplicated tests from packages, then move them to test-packages

It's my miss I have not written more ideas, and I think this is important for deciding where to have the tests in this PR.

If you look at existing tests in each packages, all `fastboot`, `fastboot-express-middleware`, `fastboot-app-server` use fixtures and implicitly rely on each other packages to build/serve the app. I think we need to move these integration pieces into `test-packages/integration-tests/test/*-test.js`, or figure out the boundary and test the unit behaviors in package's own test folder. There need to be more thoughts go into each PR when doing so.

**Comment by xg-wang** [comment]
> @xg-wang the first few of your commits show a blank diff any idea why?
> 
> [0f5b00b](https://github.com/ember-fastboot/ember-cli-fastboot/commit/0f5b00b8442bcc6c08fc2f69d4cfdb710e11eb3a)

@kiwiupover that's in the repo moved into the monorepo (for example `fastboot`), I committed a pre-merge commit to move everything to `packages/` so the later merge has no conflicts. GitHub may have some issue displaying the diff. I used the following script:

```sh
cd fastboot
mkdir -p ../packages/fastboot; mv .* * $_
mv ../packages .
mv packages/fastboot/.git .
git commit -m 'Monorepo setup: move to packages/fastboot'
# repeat for other repos
git remote add local/fastboot file://$HOME/Code/ember-fastboot/fastboot
git fetch local/fastboot
git merge local/fastboot/master --allow-unrelated-histories
```

GitHub may have some issue displaying them, you can see it's the file move by doing `git show 0f5b00b8442bcc6c08fc2f69d4cfdb710e11eb3a`

File: test-packages/custom-sandbox-app/README.md (hunk 2 above)
Comment (xg-wang) [reply test-packages/custom-sandbox-app/README.md:1]: Pushed a commit to update the readme. The difference to `basic-app` is minimal, but we still need a separate app to build. Any suggestions?

File: test-packages/hot-swap-app/README.md (hunk 3 above)
Comment (xg-wang) [reply test-packages/hot-swap-app/README.md]: Pushed a commit to update the readme

**Comment by xg-wang** [comment]
Paired with @kiwiupover, we walked through the PR and listed out the things needed:

Must:
- [x] add integration-test in ci
- [x] rename lib to test-helper
- [ ] merge into master, ask Rob to release 3.1.3-beta, confirm npm packages work

After merge:
- [ ] check contributor list is updated
- [ ] update other packages, remove files, update readme about redirect link
- [ ] over next week: transfer issues, close or re-open PRs

In follow-up PRs:
- [ ] update buildDist helper to accept prod build flag
- [ ] refactor lib/clean-dists.js to walk fs structure
- [ ] consolidate
  - consolidate hot-swap-app and custom-sandbox-app to sample-app-custom-fastboot
  - rename basic-app -> sample-app
  - Write testing-readme instruction for few things
    - `packages/` have unit tests
    - integration tests (e.g., `basic-app/test/`) move to `integration-tests`
    - Identify duplicate tests from `packages/*`
    - In the future, think about `basic-app/test/` contain app test code using `ember-cli-fastboot-testing`, to test fastboot provided test-helpers for app developers are working correctly

File: test-packages/integration-tests/package.json [hunk 9]
```diff
@@ -4,7 +4,7 @@
   "version": "3.1.2",
   "repository": "",
   "scripts": {
-    "test": "node lib/clean-dists.js && mocha"
+    "test": "node helpers/clean-dists.js && mocha"
```
Comment (kiwiupover) [inline test-packages/integration-tests/package.json:7]: +1

**Review by kiwiupover — APPROVED** [review]

**Review by kiwiupover — APPROVED** [review]
@xg-wang Great work Thomas!

**Comment by kiwiupover** [comment]
@rwjblue this is ready to be merged!

**Pull request merged**